File: src/ez_pnp2.jl
#!/usr/bin/env julia
using RobotOS
using PyCall

@pyimport moveit_commander
@pyimport rospy

@rosimport geometry_msgs.msg: PoseStamped
@rosimport ez_pick_and_place.srv: EzSceneSetup
@rosimport grasp_planning_graspit_msgs.srv: AddToDatabase, LoadDatabaseModel
@rosimport manipulation_msgs.srv: GraspPlanning

rostypegen()

using .geometry_msgs.msg
using .ez_pick_and_place.srv
using .grasp_planning_graspit_msgs.srv

planning_srv = Nothing
add_model_srv = Nothing
load_model_srv = Nothing

ez_objects = Dict()
ez_obstacles = Dict()

function validInput(req::EzSceneSetupRequest)
    return true
end

function scene_setup(req::EzSceneSetupRequest)
    res = EzSceneSetupResponse()
    if !validInput(req)
        res.success = false
        res.info = "Invalid input data"
        return res
    end
    for object in req.objects
        atd = AddToDatabaseRequest()
        atd.filename =object.file
        atd.isRobot = false
        atd.asGraspable = true
        atd.modelName = object.name
        response = add_model_srv(atd)
        objectID = response.modelID
        ez_objects[object.name] = objectID
    end
    for obstacle in req.obstacles
        atd = AddToDatabaseRequest()
        atd.filename = obstacle.file
        atd.isRobot = false
        atd.asGraspable = false
        atd.modelName = obstacle.name
        response = add_model_srv(atd)
        obstacleID = response.modelID
        ez_obstacles[obstacle.name] = obstacleID
    end
    return EzSceneSetupResponse()
end

function main()
    init_node("ez_pnp")
    scene_srv = Service("ez_pnp/scene_setup", EzSceneSetup, scene_setup)

    add_model_srv = ServiceProxy("/graspit_add_to_database", AddToDatabase)
    wait_for_service("/graspit_add_to_database")
    load_model_srv = ServiceProxy("/graspit_load_model", LoadDatabaseModel)
    wait_for_service("/graspit_load_model")
    planning_srv = ServiceProxy("/graspit_eg_planning", GraspPlanning)
    wait_for_service("/graspit_eg_planning")

    spin()
end

main()

File: src/ez_pnp2.py
#!/usr/bin/env python
import sys
import rospy
import tf2_ros
import moveit_commander

from grasp_planning_graspit_msgs.srv import AddToDatabase, LoadDatabaseModel
from ez_pick_and_place.srv import EzSceneSetup, EzStartPlanning
from moveit_msgs.srv import GraspPlanning, GetPositionIK

from ez_tools import EZToolSet

def main():
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node("ez_pnp")

    ez_tools = EZToolSet()

    ez_tools.debug = rospy.get_param("/ez_pnp/debug", False)

    ez_tools.moveit_scene = moveit_commander.PlanningSceneInterface()
    ez_tools.tf2_buffer = tf2_ros.Buffer()
    ez_tools.tf2_listener = tf2_ros.TransformListener(ez_tools.tf2_buffer)

    ez_tools.add_model_srv = rospy.ServiceProxy("/graspit_add_to_database", AddToDatabase)
    rospy.wait_for_service("/graspit_add_to_database")
    ez_tools.load_model_srv = rospy.ServiceProxy("/graspit_load_model", LoadDatabaseModel)
    rospy.wait_for_service("/graspit_load_model")
    ez_tools.planning_srv = rospy.ServiceProxy("/graspit_eg_planning", GraspPlanning)
    rospy.wait_for_service("/graspit_eg_planning")
    ez_tools.compute_ik_srv = rospy.ServiceProxy("/compute_ik", GetPositionIK)
    rospy.wait_for_service("/compute_ik")

    start_srv = rospy.Service("ez_pnp/start_planning", EzStartPlanning, ez_tools.startPlanning)
    scene_srv = rospy.Service("ez_pnp/scene_setup", EzSceneSetup, ez_tools.sceneSetup)

    rospy.spin()

main()
